contrib/lib/src/compression.rs

```diff
@@ -0,0 +1,133 @@
+//! This module provides brotli ang gzip compression for all non-image
+//! responses for requests that send Accept-Encoding br and gzip. If
+//! accepted, brotli compression is preferred over gzip.
+//!
+//! To add this feature to your Rocket application, use
+//! .attach(rocket_contrib::Compression::fairing())
+//! to your Rocket instance. Note that you must add the
+//! "compression" feature for brotli and gzip compression to your rocket_contrib
+//! dependency in Cargo.toml. Additionally, you can load only brotli compression
+//! using "brotli_compression" feature or load only gzip compression using
+//! "gzip_compression" in your rocket_contrib dependency in Cargo.toml.
+//!
+//! In the brotli algorithm, quality is set to 2 in order to have really fast
+//! compressions with compression ratio similar to gzip. Also, text and font
+//! compression mode is set regarding the Content-Type of the response.
+//!
+//! In the gzip algorithm, quality is the default (9) in order to have good
+//! compression ratio.
+//!
+//! For brotli compression, the rust-brotli crate is used.
+//! For gzip compression, flate2 crate is used.
+
+use std::io::Cursor;
+use std::io::Seek;
+use std::io::SeekFrom;
+use std::io::{Read, Write};
+
+use rocket::fairing::{Fairing, Info, Kind};
+use rocket::http::Header;
+use rocket::{Request, Response};
+
+#[cfg(feature = "brotli_compression")]
+use brotli;
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2;
+#[cfg(feature = "gzip_compression")]
+use flate2::write::GzEncoder;
+
+#[derive(Debug, Default)]
+pub struct Compression;
```

Comment:
My previous comments about streaming still stand -- were you waiting for the changes in brotli to update that code?

I have not pushed the last commit because it will not compile until the pull request in rust-brotli is merged.

There are few brotli crates in crates.io, and rust-brotli is the fastest one. A post in the Dropbox blog describes the algorithm with more details: https://blogs.dropbox.com/tech/2016/06/lossless-compression-with-brotli/.

`Default` is used in the `fairing` function, but it is not strictly necessary. `Debug` derive can also be dropped. I add the empty tuple as field in the struct like you suggest and I also remove the `derive`s. It will be pushed as soon as the pull request is merged in rust-brotli repository.